### PR TITLE
Scalarizer: Replace cl::opts with pass parameters

### PR DESCRIPTION
**Pull request by arsenm**
Preserve the existing defaults (although load-store defaulting
to false is a really bad one). Also migrate DirectX tests to new PM.

### CODE REVIEW

File: llvm/lib/Passes/PassRegistry.def [hunk 1]
```diff
@@ -573,6 +572,12 @@ FUNCTION_PASS_WITH_PARAMS(
       return StackLifetimePrinterPass(dbgs(), Type);
     },
     parseStackLifetimeOptions, "may;must")
+FUNCTION_PASS_WITH_PARAMS(
+    "scalarizer", "ScalarizerPass",
+    [](ScalarizerPassOptions Opts) { return ScalarizerPass(Opts); },
+    parseScalarizerOptions,
+    "load-store;no-load-store;variable-insert-extract;"
+    "no-variable-insert-extract;no-min-bits=N;min-bits=N;")
```
Comment (aeubanks) [inline llvm/lib/Passes/PassRegistry.def:580]: `no-min-bits` isn't a thing if I'm reading above correctly

File: llvm/test/Transforms/Scalarizer/pass-param-parse-errors.ll [hunk 2]
```diff
@@ -0,0 +1,8 @@
+; RUN: not opt -passes='scalarizer<unknown>' -disable-output %s 2>&1 | FileCheck -check-prefix=UNKNOWNERR %s
+; RUN: not opt -passes='scalarizer<;>' -disable-output %s 2>&1 | FileCheck -check-prefix=UNKNOWNERR %s
+; RUN: not opt -passes='scalarizer<min-bits=>' -disable-output %s 2>&1 | FileCheck -check-prefix=MINBITS-EMPTY-ERR %s
+; RUN: not opt -passes='scalarizer<min-bits=x>' -disable-output %s 2>&1 | FileCheck -check-prefix=MINBITS-NOTINT-ERR %s
```
Comment (aeubanks) [inline llvm/test/Transforms/Scalarizer/pass-param-parse-errors.ll:4]: add check for `no-min-bits`

**Review by aeubanks — APPROVED** [review]
lg with comments addressed

**Comment by farzonl** [comment]
We intentionally brought back the old pass manager semantics. The directX backend is still experimental so if someone removes the legacy pass manager our builds will break silently.  Our thinking was keeping around the legacy flag makes it so someone has to grep the codebase before removing.

File: llvm/lib/Passes/PassRegistry.def (hunk 1 above)
Comment (arsenm) [reply llvm/lib/Passes/PassRegistry.def:580]: I copied this from some of the other thresholds, but I don't see how it makes any sense. I think I'll just drop it

File: llvm/include/llvm/Transforms/Scalar/Scalarizer.h [hunk 3]
```diff
@@ -27,13 +27,25 @@ class Function;
...
+
+  /// Allow the scalarizer pass to scalarize insertelement/extractelement with
+  /// variable index.
+  bool ScalarizeVariableInsertExtract = true;
+
+  /// Allow the scalarizer pass to scalarize loads and store
+  ///
+  /// This is disabled by default because having separate loads and stores makes
+  /// it more likely that the -combiner-alias-analysis limits will be reached.
+  bool ScalarizeLoadStore = false;
```
Comment (farzonl) [inline llvm/include/llvm/Transforms/Scalar/Scalarizer.h:48]: This doesn't impact how we set our load/store flag so mostly fine with this change. 
https://github.com/llvm/llvm-project/blob/38a8000f30aa87a51ac0598dae3cb9166084f6a0/llvm/lib/Target/DirectX/DirectXTargetMachine.cpp#L91-L93

File: llvm/lib/Target/DirectX/DirectXTargetMachine.cpp [hunk 4]
```diff
@@ -12,6 +12,7 @@
 //===----------------------------------------------------------------------===//
 
 #include "DirectXTargetMachine.h"
+#include "DXILDataScalarization.h"
```
Comment (farzonl) [inline llvm/lib/Target/DirectX/DirectXTargetMachine.cpp:15]: Why did this need to be added? We are only using the legacy pass in this file?

**Comment by farzonl** [comment]
> > We intentionally brought back the old pass manager semantics. The directX backend is still experimental so if someone removes the legacy pass manager our builds will break silently. Our thinking was keeping around the legacy flag makes it so someone has to grep the codebase before removing.
> 
> If you want to set the options from an invocation in the backend, you can still directly pass them to the pass constructor. cl::opts are not a viable way for backends to configure passes, and only potentially useful for test flags and debugging (where running in old PM is obsolete).
Y


> > We intentionally brought back the old pass manager semantics. The directX backend is still experimental so if someone removes the legacy pass manager our builds will break silently. Our thinking was keeping around the legacy flag makes it so someone has to grep the codebase before removing.
> 
> If you want to set the options from an invocation in the backend, you can still directly pass them to the pass constructor. cl::opts are not a viable way for backends to configure passes, and only potentially useful for test flags and debugging (where running in old PM is obsolete).

We weren't using it to configure the backend just as a hint/marker that the legacy pass was still in use without folks having to build the DirectX backend. In any case I don't have a strong opinion on this so I'm fine switching it out.

You will however want to remove `scalarizer` from the optdriver:  
https://github.com/llvm/llvm-project/blob/ce72c76e9bd0bf74af614cae6a9f85cfd4720e95/llvm/tools/opt/optdriver.cpp#L378

**Comment by llvm-beanz** [comment]
Generally `cl::opt` values are static initializers, which cause performance degradation and other problems. Moving away from `cl::opt` is definitely a net win.

**Review by farzonl — APPROVED** [review]

File: llvm/lib/Target/DirectX/DirectXTargetMachine.cpp (hunk 4 above)
Comment (arsenm) [reply llvm/lib/Target/DirectX/DirectXTargetMachine.cpp:15]: Need the pass constructor, transitively included by DirectXPassRegistry.def

**Pull request merged**